### PR TITLE
[SPARK-43781][SQL] Fix IllegalStateException when cogrouping two datasets derived from the same source

### PR DESCRIPTION
**Pull request by Hisoka-X**
### What changes were proposed in this pull request?
When cogroup two datasets derived from same source, eg:
```scala
val inputType = StructType(Array(StructField("id", LongType, false),
   StructField("type", StringType, false)))
val keyType = StructType(Array(StructField("id", LongType, false)))

val inputRows = new java.util.ArrayList[Row]()
inputRows.add(Row(1L, "foo"))
inputRows.add(Row(1L, "bar"))
inputRows.add(Row(2L, "foo"))
val input = spark.createDataFrame(inputRows, inputType)
val fooGroups = input.filter("type = 'foo'").groupBy("id").as(RowEncoder(keyType),
   RowEncoder(inputType))
val barGroups = input.filter("type = 'bar'").groupBy("id").as(RowEncoder(keyType),
   RowEncoder(inputType))

val result = fooGroups.cogroup(barGroups) { case (row, iterator, iterator1) =>
   iterator.toSeq ++ iterator1.toSeq
}(RowEncoder(inputType)).collect()
```
The error will be reported:
```
21:03:27.651 ERROR org.apache.spark.executor.Executor: Exception in task 1.0 in stage 0.0 (TID 1)
java.lang.IllegalStateException: Couldn't find id#19L in [id#0L,type#1]
	at org.apache.spark.sql.catalyst.expressions.BindReferences$$anonfun$bindReference$1.applyOrElse(BoundAttribute.scala:80)
	at org.apache.spark.sql.catalyst.expressions.BindReferences$$anonfun$bindReference$1.applyOrElse(BoundAttribute.scala:73)
	...
```
The reason are `DeduplicateRelations` rewrite `LocalRelation` but can't rewrite `left(right)Group` and `left(right)Attr` in `CoGroup`. In fact, the `Join` will face same situation. But `Join` regenerate plan when invoke itself to avoid this situation. Please refer https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala#L1089

This PR let `DeduplicateRelations` handle with `CoGroup` case

### Why are the changes needed?
Fix IllegalStateException when cogrouping two datasets derived from the same source

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Add new test

### CODE REVIEW

**Comment by cloud-fan** [comment]
can we make DeduplicateRelations handle `CoGroup`?

**Comment by Hisoka-X** [comment]
> can we make DeduplicateRelations handle `CoGroup`?

Ok for me, before do this, please check #41347

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/DeduplicateRelations.scala [hunk 1]
```diff
@@ -220,7 +220,52 @@ object DeduplicateRelations extends Rule[LogicalPlan] {
...
+                  attr.copy(child = canRewriteAttrs(attr.child.asInstanceOf[Attribute]))
+                } else {
+                  attr
+                }
+              })
+            }
+
+            planWithNewChildren match {
+              case c @ CoGroup(_, keyDeserializer, leftDeserializer, rightDeserializer,
+              leftGroup, rightGroup, leftAttr, rightAttr, leftOrder, rightOrder, _, left,
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/DeduplicateRelations.scala:247]: nit: let's not use pattern match if there are too many parameters. `case c: CoGroup` is better.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/DeduplicateRelations.scala [hunk 2]
```diff
@@ -220,7 +220,52 @@ object DeduplicateRelations extends Rule[LogicalPlan] {
...
+              right) =>
+                // SPARK-43781: CoGroup is a special case, as it has different output attributes
+                // from its children. We need to update the output attributes of CoGroup manually.
+                val newLeftAttr = rewriteAttrsMatchWithSubPlan(leftAttr, attrMap, left.output)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/DeduplicateRelations.scala:251]: The code looks a bit weird. `leftAttr` must come from `left.output`, so the code should simply replace attributes in an `Expression` using `attrMap`. e.g.
```
leftAttr.map(_.transform {
  case a: AttributeReference => updateAttr(a, attrMap)
}.asInstanceOf[Attribute])
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/DeduplicateRelations.scala [hunk 3]
```diff
@@ -220,7 +220,51 @@ object DeduplicateRelations extends Rule[LogicalPlan] {
...
+              val canRewriteAttrs = attrMap.filter(a => planOutput.contains(a._2))
+              attrs.map(attr => canRewriteAttrs.getOrElse(attr, attr))
+            }
+
+            def rewriteOrderMatchWithSubPlan(
+                attrs: Seq[SortOrder],
+                attrMap: AttributeMap[Attribute],
+                planOutput: Seq[Attribute]): Seq[SortOrder] = {
+              val canRewriteAttrs = attrMap.filter(a => planOutput.contains(a._2))
+              attrs.filter(_.child.isInstanceOf[Attribute]).map(attr => {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/DeduplicateRelations.scala:236]: why do we only handle top-level attributes? how about something like `a + 1`? Shall we use transform to rewrite all attributes in the ordering expression?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/DeduplicateRelations.scala [hunk 4]
```diff
@@ -220,7 +220,51 @@ object DeduplicateRelations extends Rule[LogicalPlan] {
...
+                  attr
+                }
+              })
+            }
+
+            planWithNewChildren match {
+              case c: CoGroup =>
+                // SPARK-43781: CoGroup is a special case, as it has different output attributes
+                // from its children. We need to update the output attributes of CoGroup manually.
+                val newLeftAttr = c.leftAttr.map(attr => attrMap.getOrElse(attr, attr))
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/DeduplicateRelations.scala:249]: why the left-right attributes does not need to call `rewriteAttrsMatchWithSubPlan`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/DeduplicateRelations.scala [hunk 5]
```diff
@@ -220,7 +220,51 @@ object DeduplicateRelations extends Rule[LogicalPlan] {
...
+                val newLeftGroup = rewriteAttrsMatchWithSubPlan(c.leftGroup, attrMap, c.left.output)
+                c.copy(
+                  keyDeserializer = c.keyDeserializer.asInstanceOf[UnresolvedDeserializer]
+                    .copy(inputAttributes = newLeftGroup),
+                  leftDeserializer = c.leftDeserializer.asInstanceOf[UnresolvedDeserializer]
+                    .copy(inputAttributes = newLeftAttr),
+                  rightDeserializer = c.rightDeserializer.asInstanceOf[UnresolvedDeserializer]
+                    .copy(inputAttributes = newRightAttr),
+                  leftGroup = newLeftGroup,
+                  rightGroup = rewriteAttrsMatchWithSubPlan(c.rightGroup, attrMap, c.right.output),
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/DeduplicateRelations.scala:261]: can we use a consistent code style? either we put all these things in local variables, or inline all the them.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/DeduplicateRelations.scala [hunk 6]
```diff
@@ -220,7 +220,54 @@ object DeduplicateRelations extends Rule[LogicalPlan] {
...
+                  case a: AttributeReference =>
+                    canRewriteAttrs.getOrElse(a, a)
+                }
+                attr
+              })
+            }
+
+            planWithNewChildren match {
+              case c: CoGroup =>
+                // SPARK-43781: CoGroup is a special case, as it has different output attributes
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/DeduplicateRelations.scala:247]: What do you mean by "it has different output attributes"?

There are similar cases where it creates new output attributes (not from the child):  `MapPartitionsInR`, `MapPartitionsInRWithArrow`, `MapElements`, `MapGroups`, `FlatMapGroupsWithState`, `FlatMapGroupsInR`, `FlatMapGroupsInR`, `FlatMapGroupsInRWithArrow`, `FlatMapGroupsInPandas`,  `MapInPandas`, `PythonMapInArrow`,  and `FlatMapGroupsInPandasWithState`.

There's another CoGroup operation in Python (`cogroup.apply()` that is `FlatMapCoGroupsInPandas`). Would be great to double check.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/DeduplicateRelations.scala [hunk 7]
```diff
@@ -220,7 +220,55 @@ object DeduplicateRelations extends Rule[LogicalPlan] {
...
+                attrs: Seq[SortOrder],
+                attrMap: AttributeMap[Attribute],
+                planOutput: Seq[Attribute]): Seq[SortOrder] = {
+              val canRewriteAttrs = attrMap.filter(a => planOutput.contains(a._2))
+              attrs.map(attr => {
+                attr.transformWithPruning(_.containsPattern(ATTRIBUTE_REFERENCE)) {
+                  case a: AttributeReference =>
+                    canRewriteAttrs.getOrElse(a, a)
+                }
+                attr
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/DeduplicateRelations.scala:241]: why return the input attr? This means the rewrite is completely ignored.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/DeduplicateRelations.scala [hunk 8]
```diff
@@ -220,7 +220,42 @@ object DeduplicateRelations extends Rule[LogicalPlan] {
           if (attrMap.isEmpty) {
             planWithNewChildren
           } else {
-            planWithNewChildren.rewriteAttrs(attrMap)
+            def rewriteAttrsMatchWithSubPlan[T <: Expression](
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/DeduplicateRelations.scala:223]:
```suggestion
            def rewriteAttrs[T <: Expression](
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/DeduplicateRelations.scala [hunk 9]
```diff
@@ -220,7 +220,42 @@ object DeduplicateRelations extends Rule[LogicalPlan] {
...
+                attrs: Seq[T],
+                attrMap: Map[Attribute, Attribute]): Seq[T] = {
+              attrs.map(attr => {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/DeduplicateRelations.scala:226]: code style:
```
...map { expr =>
  expr.transform...
}
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/DeduplicateRelations.scala (hunk 4 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/DeduplicateRelations.scala:241]: should the same apple to `newRightAttr`?

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/DeduplicateRelations.scala:241]: and to left/right group?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/DeduplicateRelations.scala (hunk 6 above)
Comment (HyukjinKwon) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/DeduplicateRelations.scala]: Yup, that's fine to do them as a followup.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/DeduplicateRelations.scala [hunk 10]
```diff
@@ -220,7 +220,42 @@ object DeduplicateRelations extends Rule[LogicalPlan] {
           if (attrMap.isEmpty) {
             planWithNewChildren
           } else {
-            planWithNewChildren.rewriteAttrs(attrMap)
+            def rewriteAttrs[T <: Expression](
+                exprs: Seq[T],
+                attrMap: Map[Attribute, Attribute]): Seq[T] = {
+              exprs.map(expr => {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/DeduplicateRelations.scala:226]: https://github.com/apache/spark/pull/41554#discussion_r1288731966

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/DeduplicateRelations.scala [hunk 11]
```diff
@@ -220,7 +220,41 @@ object DeduplicateRelations extends Rule[LogicalPlan] {
           if (attrMap.isEmpty) {
             planWithNewChildren
           } else {
-            planWithNewChildren.rewriteAttrs(attrMap)
+            def rewriteAttrs[T <: Expression](
+                exprs: Seq[T],
+                attrMap: Map[Attribute, Attribute]): Seq[T] = {
+              exprs.map(expr =>
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/DeduplicateRelations.scala:226]:
```suggestion
              exprs.map { expr =>
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/DeduplicateRelations.scala [hunk 12]
```diff
@@ -220,7 +220,42 @@ object DeduplicateRelations extends Rule[LogicalPlan] {
...
             planWithNewChildren
           } else {
-            planWithNewChildren.rewriteAttrs(attrMap)
+            def rewriteAttrs[T <: Expression](
+                exprs: Seq[T],
+                attrMap: Map[Attribute, Attribute]): Seq[T] = {
+              exprs.map { expr =>
+                expr.transformWithPruning(_.containsPattern(ATTRIBUTE_REFERENCE)) {
+                  case a: AttributeReference =>
+                    attrMap.getOrElse(a, a)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/DeduplicateRelations.scala:229]:
```suggestion
                  case a: AttributeReference => attrMap.getOrElse(a, a)
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/DeduplicateRelations.scala [hunk 13]
```diff
@@ -220,7 +220,42 @@ object DeduplicateRelations extends Rule[LogicalPlan] {
...
+                }.asInstanceOf[T]
+              }
+            }
+
+            planWithNewChildren match {
+              case c: CoGroup =>
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/DeduplicateRelations.scala:235]: Can we add a TODO to handle other similar plans?

**Review by cloud-fan — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/DeduplicateRelations.scala (hunk 13 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/DeduplicateRelations.scala:235]:
```
// TODO (SPARK-XXX): ...
```

**Comment by cloud-fan** [comment]
thanks, merging to master/3.5!

**Pull request closed without merging**